packages/flutter_test/lib/src/_goldens_io.dart

```diff
@@ -0,0 +1,208 @@
+    : basedir = _getBasedir(testFile, pathStyle),
+      _path = _getPath(pathStyle);
+
+  static path.Context _getPath(path.Style style) {
+    return path.Context(style: style ?? path.Style.platform);
+  }
+
+  static Uri _getBasedir(Uri testFile, path.Style pathStyle) {
+    final path.Context context = _getPath(pathStyle);
+    final String testFilePath = context.fromUri(testFile);
+    final String testDirectoryPath = context.dirname(testFilePath);
+    return context.toUri(testDirectoryPath + context.separator);
+  }
+
+  /// The directory in which the test was loaded.
+  ///
+  /// Golden file keys will be interpreted as file paths relative to this
+  /// directory.
+  final Uri basedir;
+
+  /// Path context exists as an instance variable rather than just using the
+  /// system path context in order to support testing, where we can spoof the
+  /// platform to test behaviors with arbitrary path styles.
+  final path.Context _path;
+
+  @override
+  Future<bool> compare(Uint8List imageBytes, Uri golden) async {
+    final File goldenFile = _getGoldenFile(golden);
+    if (!goldenFile.existsSync()) {
+      throw test_package.TestFailure('Could not be compared against non-existent file: "$golden"');
+    }
+    final List<int> goldenBytes = await goldenFile.readAsBytes();
+    final ComparisonResult result = GoldenFileComparator.compareLists(imageBytes, goldenBytes);
+
+    if (!result.passed) {
+      String additionalFeedback = '';
+      if (result.diffs != null) {
+        additionalFeedback = '\nFailure feedback can be found at ${path.join(basedir.path, 'failures')}';
+        final Map<String, Object> diffs = result.diffs;
+        diffs.forEach((String name, Object untypedImage) {
+          final Image image = untypedImage;
+          final File output = _getFailureFile(name, golden);
+          output.parent.createSync(recursive: true);
+          output.writeAsBytesSync(encodePng(image));
+        });
+      }
+      throw test_package.TestFailure('Golden "$golden": ${result.error}$additionalFeedback');
+    }
+    return result.passed;
+  }
+
+  @override
+  Future<void> update(Uri golden, Uint8List imageBytes) async {
+    final File goldenFile = _getGoldenFile(golden);
+    await goldenFile.parent.create(recursive: true);
+    await goldenFile.writeAsBytes(imageBytes, flush: true);
+  }
+
+  File _getGoldenFile(Uri golden) {
+    return File(_path.join(_path.fromUri(basedir), _path.fromUri(golden.path)));
+  }
+
+  File _getFailureFile(String failure, Uri golden) {
+    final String fileName = golden.pathSegments[0];
+    final String testName = fileName.split(path.extension(fileName))[0]
+      + '_'
+      + failure
+      + '.png';
+    return File(_path.join('failures', testName));
+  }
+}
+
+/// Returns a [ComparisonResult] to describe the pixel differential of the
+/// [test] and [master] image bytes provided.
+ComparisonResult compareLists(List<int> test, List<int> master) {
```

Comment:
I haven't worked out all the details, but I suspect that with sufficient cleverness the analyzer and/or kernel compiler can be made to check that two different libraries implement the same signature by reducing it to the problem of checking that a class correctly implements an interface and then injecting the class into the right place, like we do in the tool. In a world where we want to do something like that, top-levels and statics would get in the way a bit.

Reply:
I've implemented this check manually in the engine repo by walking both libraries are comparing types, its not great.

Currently we are running tests in both environments so we can catch type errors in CI.

Longer term I would like to not use config specific imports at all, but that is somewhat gated on what is decided on the SDK structure for flutter for web.